### DISCUSSION TITLE
Supplying Values From Request for Webhook Body and Headers

### DISCUSSION THREAD

**Discussion by mathgr22** [wiremock/WireMock.Net#1195] [Q&A]
I'm trying to determine if the functionality I'm trying to achieve is possible or if it is not currently supported. 

Basically what I'm trying to accomplish is after a response is sent as a result of an HTTP post, I want to send one or more `Webhook` responses to a different URL than the caller to mock a 3rd-party service we use. I want the data sent in the body and headers to replicate data that was sent via the initial request. This can be done effectively via the Handlebars transformer, however the problem is I need to hash the body of the `Webhook` via HMAC and supply a signature in the header. As far as I am aware, Handlebars is not sophisticated enough to do this. 

Essentially I want to accomplish this via having a new Webhook request created each time there is a request as can be accomplished via the `IResponseBuilder.WithBody()` overload that allows a function to be supplied that will be called each time a request goes through. There doesn't seem to be the ability to do this with the `IResponseWithAProvider.WithWebhook()` method. 

Is the functionality I'm trying to implement not currently possible, or am I just not understanding how to do it?

**Comment by StefH** [comment]
The WebHook functionality is not that extended that you can provide your own code for calculating something that HMAC.

I can take a look what it takes to add this, but it involves a lot of changes I think.

---

Maybe you can write your own handlebars function to create that HMAC?

**Reply by mathgr22** [reply]
Thanks for the quick reply. I'm not sure writing a Handlebars function will work, because the headers will then need access to the body, and my understanding is that they are transformed separately.

**Reply by StefH** [reply]
I think all transformers have access to [TransformModel](https://github.com/WireMock-Net/WireMock.Net/blob/master/src/WireMock.Net/Transformers/TransformModel.cs) which does include the request-body

**Reply by mathgr22** [reply]
That does appear to be the case, however, I'm wanting to transform the body first, and subsequently transform the headers based on the result of the transformed body. I'm looking at the code in `WebhookSender.SendAsync()` (assuming this is the correct place) and it seems it transforms the header with the pre-transformed response as far as I can tell.

**Reply by StefH** [reply]
You are correct. This will not work. Then we need WithBody() overload that allows a function.

**Reply by StefH** [reply]
Another option would be to add a delegate to the WebhookSettings which allows you to access the transformed body + headers.
And you can still modify data if you want.

``` c#
public class WebhookSettings : HttpClientSettings
{
  public delegate void PostTransform(IMapping mapping, IBodyData? body, IDictionary<string, WireMockList<string>>? headers);
}
```

**Reply by mathgr22** [reply]
This was an alternate solution I had considered, I think it would work for my use case, but I haven't tested it yet.

**Reply by StefH** [reply]
https://github.com/WireMock-Net/WireMock.Net/pull/1223
